mincall/input_readers.py

```diff
@@ -245,6 +250,88 @@ def get_basecalled_data(self, h5, target_read, sampling_rate):
+        return (signal - 646.11133) / 75.673653
+
+    def get_signal(self, fast5_path):
+        with h5py.File(fast5_path, 'r') as h5:
+            reads = h5['Raw/Reads']
+            target_read = list(reads.keys())[0]
+            h5_logits = h5['Analyses/MinCall/Logits']
+            shrink_factor = h5_logits.attrs.get("shrink_factor", 8)  # Default 8..TODO
+            signal_len = h5_logits.shape[0] * shrink_factor
+            signal = h5['Raw/Reads/' + target_read]['Signal']
+            start_pad = h5_logits.attrs['start_pad']
+            signal = signal[start_pad:start_pad + signal_len]
+
+        return self.preprocessSignal(signal).reshape(1, -1, 1), start_pad
+
+    def read_fast5_raw_ref(self, fast5_path, ref_path, block_size_x, block_size_y, num_blocks, n_samples_per_ref=1,
+                           verify_file=True):
+
+        with h5py.File(fast5_path, 'r') as h5:
+            reads = h5['Raw/Reads']
+            target_read = list(reads.keys())[0]
+            try:
+                h5_logits = h5['Analyses/MinCall/Logits']
+            except:
+                raise MissingMincallLogits()
+
+            try:
+                h5_refalignment = h5['Analyses/MinCall/RefAlignment']
+            except:
+                raise MissingMincallAlignedRef()
+
+            shrink_factor = h5_logits.attrs.get("shrink_factor", 8)  # Default 8..TODO
+
+            signal = h5['Raw/Reads/' + target_read]['Signal']
+            start_pad = h5_logits.attrs['start_pad']
+            signal_len = h5_logits.shape[0] * shrink_factor
+            signal = signal[start_pad:start_pad + signal_len]
+            signal = self.preprocessSignal(signal)
+
+            # -2 for first and last that are always skipped
+            num_blocks_max = h5_logits.shape[0] * shrink_factor // block_size_x - 2
+            if num_blocks_max < num_blocks:
+                raise InsufficientDataBlocks("Has only {} blocks, while requesting {} + first and last"
+                                             .format(num_blocks_max, num_blocks))
+
+            n_different_blocks = num_blocks_max - num_blocks + 1
+            n_different_non_overlap_blocks = n_different_blocks // num_blocks
+            n_samples = min(n_samples_per_ref, n_different_non_overlap_blocks)
+
+            for i in range(n_samples):
+                x = np.zeros([block_size_x * num_blocks, 1], dtype=np.float32)
+                x_len = min(len(signal), block_size_x * num_blocks)
+
+                # start from i-th block
+                # does not guarantee non overlapping blocks but ok enough
+                start_block = randint(1, n_different_blocks)
+                x_offset = start_block * block_size_x
+
+                x[:x_len, 0] = signal[x_offset:x_offset + x_len]
+
+                y = np.full([block_size_y * num_blocks], 9, dtype=np.uint8)
+                y_len = np.zeros([num_blocks], dtype=np.int32)
+
+                for i in range(num_blocks):
```

Comment:
iterator "i" shadows outer iterator (even though outer "i" is not actually used)